File: MatlabAPI/private/maskApi.cpp
/**************************************************************************
* Microsoft COCO Toolbox.      Version 1.0
* Data, paper, and tutorials available at:  http://mscoco.org/
* Code written by Piotr Dollar and Tsung-Yi Lin, 2015.
* Licensed under the Simplified BSD License [see coco/license.txt]
**************************************************************************/
#include "maskApi.h"
#include <math.h>
#include <stdlib.h>
#include <algorithm>

using namespace std;

void RLE::encode( const byte *mask, siz h0, siz w0 ) {
  h=h0; w=w0; counts.clear(); uint c=0; byte p=0;
  for( siz i=0; i<w*h; i++ ) {
    if(mask[i]!=p) { counts.push_back(c); c=0; p=mask[i]; } c++; }
  counts.push_back(c);
}

void RLE::decode( byte *mask ) const {
  byte v=0; for( siz i=0; i<counts.size(); i++ ) {
    for( siz j=0; j<counts[i]; j++ ) *(mask++)=v; v=!v; }
}

void RLE::merge( const RLE &A, const RLE &B, bool intersect ) {
  counts.clear(); w=A.w; h=B.h;
  if( A.h!=B.h || A.w!=B.w ) { w=h=0; return; }
  siz ka, kb, a, b; uint c, ca, cb, cc, ct; bool v, va, vb, vp;
  ca=A.counts[0]; ka=A.counts.size(); v=va=vb=0;
  cb=B.counts[0]; kb=B.counts.size(); a=b=1; cc=0; ct=1;
  while( ct>0 ) {
    c=min(ca,cb); cc+=c; ct=0;
    ca-=c; if(!ca && a<ka) { ca=A.counts[a++]; va=!va; } ct+=ca;
    cb-=c; if(!cb && b<kb) { cb=B.counts[b++]; vb=!vb; } ct+=cb;
    vp=v; if(intersect) v=va&&vb; else v=va||vb;
    if( v!=vp||ct==0 ) { counts.push_back(cc); cc=0; }
  }
}

void RLE::merge( const RLES &Rs, bool intersect ) {
  counts.clear(); w=h=0; siz i, n=Rs.size(); if(n==0) return;
  for(i=1; i<n; i++) if(Rs[i].h!=Rs[0].h||Rs[i].w!=Rs[0].w) return;
  if(n==1) *this=Rs[0]; else merge(Rs[0],Rs[1],intersect);
  for(i=2; i<n; i++) { RLE T=*this; merge(T,Rs[i],intersect); }
}

double RLE::iou( RLE &dt, RLE &gt, byte iscrowd ) {
  if( dt.h!=gt.h || dt.w!=gt.w ) return -1;
  siz ka, kb, a, b; uint c, ca, cb, ct, i, u; bool va, vb;
  ca=dt.counts[0]; ka=dt.counts.size(); va=vb=0;
  cb=gt.counts[0]; kb=gt.counts.size(); a=b=1; i=u=0; ct=1;
  while( ct>0 ) {
    c=min(ca,cb); if(va||vb) { u+=c; if(va&&vb) i+=c; } ct=0;
    ca-=c; if(!ca && a<ka) { ca=dt.counts[a++]; va=!va; } ct+=ca;
    cb-=c; if(!cb && b<kb) { cb=gt.counts[b++]; vb=!vb; } ct+=cb;
  }
  if(i==0) return 0; if(iscrowd) u=dt.area();
  return double(i)/double(u);
}

void RLE::iou( RLES &dt, RLES &gt, byte *iscrowd, double *o ) {
  siz m=dt.size(), n=gt.size(), g, d;
  BB db=new double[4*m], gb=new double[4*n];
  for( d=0; d<m; d++ ) dt[d].toBbox(db+d*4);
  for( g=0; g<n; g++ ) gt[g].toBbox(gb+g*4);
  RLE::iou( db, gb, m, n, iscrowd, o );
  for( g=0; g<n; g++ ) for( d=0; d<m; d++ ) if(o[g*m+d]>0) {
    o[g*m+d]=RLE::iou(dt[d],gt[g],iscrowd?iscrowd[g]:0); }
  delete [] db; delete [] gb;
}

void RLE::iou( BB dt, BB gt, siz m, siz n, byte *iscrowd, double *o ) {
  double h, w, i, u, ga, da; siz g, d; bool crowd;
  for( g=0; g<n; g++ ) {
    BB G=gt+g*4; ga=G[2]*G[3]; crowd=iscrowd!=NULL && iscrowd[g];
    for( d=0; d<m; d++ ) {
      BB D=dt+d*4; da=D[2]*D[3]; o[g*m+d]=0;
      w=min(D[2]+D[0],G[2]+G[0])-max(D[0],G[0]); if(w<=0) continue;
      h=min(D[3]+D[1],G[3]+G[1])-max(D[1],G[1]); if(h<=0) continue;
      i=w*h; u = crowd ? da : da+ga-i; o[g*m+d]=i/u;
    }
  }
}

uint RLE::area() const {
  uint a=0; siz i, k=counts.size();
  for(i=1; i<k; i+=2) a+=counts[i]; return a;
}

void RLE::toBbox( BB bbox ) const {
  uint hu=uint(h); siz k=counts.size(); k=siz(k/2)*2;
  if(k==0) { bbox[0]=bbox[1]=bbox[2]=bbox[3]=0; return; }
  vector<uint> cc, y, x; cc=counts; x.resize(k); y.resize(k);
  for(siz i=1; i<k; i++) cc[i]+=cc[i-1];
  for(siz i=1; i<k; i+=2) cc[i]--;
  for(siz i=0; i<k; i++) { y[i]=cc[i]%hu; x[i]=(cc[i]-y[i])/hu; }
  bbox[0] = *min_element(x.begin(),x.end());
  bbox[1] = *min_element(y.begin(),y.end());
  bbox[2] = *max_element(x.begin(),x.end()) - bbox[0] + 1;
  bbox[3] = *max_element(y.begin(),y.end()) - bbox[1] + 1;
}

void RLE::frBbox( const BB bbox, siz h0, siz w0 ) {
  double xs=bbox[0], xe=bbox[0]+bbox[2];
  double ys=bbox[1], ye=bbox[1]+bbox[3];
  double x[4] = {xs,xs,xe,xe}, y[4]={ys,ye,ye,ys};
  frPoly( x, y, 4, h0, w0 );
}

void RLE::frPoly( double *x0, double *y0, siz k, siz h0, siz w0 ) {
  h=h0; w=w0; counts.clear(); siz i; double scale=5;
  // upsample and get discrete points densely along entire boundary
  vector<int> x1(k), y1(k), x, y;
  for(i=0; i<k; i++) x1[i]=int(scale*x0[i]+.5);
  for(i=0; i<k; i++) y1[i]=int(scale*y0[i]+.5);
  if( x1[k-1]!=x1[0] || y1[k-1]!=y1[0] ) {
    x1.push_back(x1[0]); y1.push_back(y1[0]); k++; }
  for( i=0; i<k-1; i++ ) {
    int xs=x1[i], xe=x1[i+1], ys=y1[i], ye=y1[i+1], dx, dy, t;
    dx=abs(xe-xs); dy=abs(ys-ye); bool flip; double s;
    if(!dx&&!dy) { x.push_back(xs); y.push_back(ys); continue; }
    flip = (dx>=dy && xs>xe) || (dx<dy && ys>ye);
    if(flip) { t=xs; xs=xe; xe=t; t=ys; ys=ye; ye=t; }
    s = dx>=dy ? double(ye-ys)/dx : double(xe-xs)/dy;
    if(dx>=dy) for( int j=0; j<=dx; j++ ) {
      t=flip?dx-j:j; x.push_back(t+xs); y.push_back(int(ys+s*t+.5));
    } else for( int j=0; j<=dy; j++ ) {
      t=flip?dy-j:j; y.push_back(t+ys); x.push_back(int(xs+s*t+.5));
    }
  }
  // get points along y-boundary and downsample
  x1.clear(); y1.clear(); k=x.size(); double xd, yd;
  for( i=1; i<k; i++ ) if(x[i]!=x[i-1]) {
    xd=double(x[i]<x[i-1]?x[i]:x[i]-1); xd=(xd+.5)/scale-.5;
    if( floor(xd)!=xd || xd<0 || xd>w-1 ) continue;
    yd=double(y[i]<y[i-1]?y[i]:y[i-1]); yd=(yd+.5)/scale-.5;
    if(yd<0) yd=0; else if(yd>h) yd=h; yd=ceil(yd);
    x1.push_back(int(xd)); y1.push_back(int(yd));
  }
  // compute rle encoding given y-boundary points
  k=x1.size(); x.resize(k);
  for( i=0; i<k; i++ ) x[i]=x1[i]*int(h)+y1[i];
  sort(x.begin(),x.end()); int p=0; x.push_back(int(h*w)); k++;
  for( i=0; i<k; i++ ) { counts.push_back(uint(x[i]-p)); p=x[i]; }
  vector<uint> c(counts); counts.clear(); counts.push_back(c[0]); i=1;
  while(i<k) if(c[i]>0) counts.push_back(c[i++]); else {
    i++; if(i<k) counts.back()+=c[i++]; }
}

void RLE::toString( string &s ) const {
  // Similar to LEB128 but using 6 bits/char and ascii chars 48-111.
  siz i, n=counts.size(); s.clear(); long x; bool more;
  for( i=0; i<n; i++ ) {
    x=long(counts[i]); if(i>2) x-=long(counts[i-2]); more=true;
    while( more ) {
      char c=x & 0x1f; x >>= 5; more=(c & 0x10) ? x!=-1 : x!=0;
      if(more) c |= 0x20; c+=48; s.push_back(c);
    }
  }
}

void RLE::frString( const string &s, siz h0, siz w0 ) {
  h=h0; w=w0; counts.clear(); siz i=0, k=0, m; long x; bool more;
  while( s[k] ) {
    x=0; m=0; more=true;
    while( more ) {
      char c=s[k]-48; x |= (c & 0x1f) << 5*m;
      more = c & 0x20; k++; m++;
      if(!more && (c & 0x10)) x |= -1 << 5*m;
    }
    if(i>2) x+=long(counts[i-2]); i++; counts.push_back(uint(x));
  }
}

File: LuaAPI/MaskApi.lua
--[[----------------------------------------------------------------------------

Interface for manipulating masks stored in RLE format.

For an overview of RLE please see http://mscoco.org/dataset/#download.
Additionally, more detailed information can be found in the Matlab MaskApi.m:
  https://github.com/pdollar/coco/blob/master/MatlabAPI/MaskApi.m

The following API functions are defined:
  encode - Encode binary masks using RLE.
  decode - Decode binary masks encoded via RLE.
  merge  - Compute union or intersection of encoded masks.
  iou    - Compute intersection over union between masks.
  area   - Compute area of encoded masks.
  toBbox - Get bounding boxes surrounding encoded masks.
  frBbox - Convert bounding boxes to encoded masks.
  frPoly - Convert polygon to encoded mask.

Usage:
  Rs     = MaskApi.encode( masks )
  masks  = MaskApi.decode( Rs )
  R      = MaskApi.merge( Rs, [intersect=false] )
  o      = MaskApi.iou( dt, gt, [iscrowd=false] )
  a      = MaskApi.area( Rs )
  bbs    = MaskApi.toBbox( Rs )
  Rs     = MaskApi.frBbox( bbs, h, w )
  R      = MaskApi.frPoly( poly, h, w )
  img    = MaskApi.drawMasks( img, masks, [maxn=n], [alpha=.4], [clrs] )
For detailed usage information please see cocoDemo.lua (coming soon).

In the API the following formats are used:
  R,Rs   - [table] Run-length encoding of binary mask(s)
  masks  - [nxhxw] Binary mask(s)
  bbs    - [nx4] Bounding box(es) stored as [x y w h]
  poly   - Polygon stored as {[x1 y1 x2 y2...],[x1 y1 ...],...}
  dt,gt  - May be either bounding boxes or encoded masks
Both poly and bbs are 0-indexed (bbox=[0 0 1 1] encloses first pixel).

Common Objects in COntext (COCO) Toolbox.      version 3.0
Data, paper, and tutorials available at:  http://mscoco.org/
Code written by Pedro O. Pinheiro and Piotr Dollar, 2016.
Licensed under the Simplified BSD License [see coco/license.txt]

------------------------------------------------------------------------------]]

local ffi = require 'ffi'
local coco = require 'coco.env'

coco.MaskApi = {}
local MaskApi = coco.MaskApi

coco.libmaskapi = ffi.load(package.searchpath('libmaskapi',package.cpath))
local libmaskapi = coco.libmaskapi

--------------------------------------------------------------------------------

MaskApi.encode = function( masks )
  local n, h, w = masks:size(1), masks:size(2), masks:size(3)
  masks = masks:type('torch.ByteTensor'):transpose(2,3)
  local data = masks:contiguous():data()
  local Qs = MaskApi._rlesInit(n)
  libmaskapi.rleEncode(Qs[0],data,h,w,n)
  return MaskApi._rlesToLua(Qs,n)
end

MaskApi.decode = function( Rs )
  local Qs, n, h, w = MaskApi._rlesFrLua(Rs)
  local masks = torch.ByteTensor(n,w,h):zero():contiguous()
  libmaskapi.rleDecode(Qs,masks:data(),n)
  MaskApi._rlesFree(Qs,n)
  return masks:transpose(2,3)
end

MaskApi.merge = function( Rs, intersect )
  intersect = intersect or 0
  local Qs, n, h, w = MaskApi._rlesFrLua(Rs)
  local Q = MaskApi._rlesInit(1)
  libmaskapi.rleMerge(Qs,Q,n,intersect)
  MaskApi._rlesFree(Qs,n)
  return MaskApi._rlesToLua(Q,1)[1]
end

MaskApi.iou = function( dt, gt, iscrowd )
  if not iscrowd then iscrowd = NULL else
    iscrowd = iscrowd:type('torch.ByteTensor'):contiguous():data()
  end
  if torch.isTensor(gt) and torch.isTensor(dt) then
    local nDt, k = dt:size(1), dt:size(2); assert(k==4)
    local nGt, k = gt:size(1), gt:size(2); assert(k==4)
    local dDt = dt:type('torch.DoubleTensor'):contiguous():data()
    local dGt = gt:type('torch.DoubleTensor'):contiguous():data()
    local o = torch.DoubleTensor(nGt,nDt):contiguous()
    libmaskapi.bbIou(dDt,dGt,nDt,nGt,iscrowd,o:data())
    return o:transpose(1,2)
  else
    local qDt, nDt = MaskApi._rlesFrLua(dt)
    local qGt, nGt = MaskApi._rlesFrLua(gt)
    local o = torch.DoubleTensor(nGt,nDt):contiguous()
    libmaskapi.rleIou(qDt,qGt,nDt,nGt,iscrowd,o:data())
    MaskApi._rlesFree(qDt,nDt); MaskApi._rlesFree(qGt,nGt)
    return o:transpose(1,2)
  end
end

MaskApi.area = function( Rs )
  local Qs, n, h, w = MaskApi._rlesFrLua(Rs)
  local a = torch.IntTensor(n):contiguous()
  libmaskapi.rleArea(Qs,n,a:data())
  MaskApi._rlesFree(Qs,n)
  return a
end

MaskApi.toBbox = function( Rs )
  local Qs, n, h, w = MaskApi._rlesFrLua(Rs)
  local bb = torch.DoubleTensor(n,4):contiguous()
  libmaskapi.rleToBbox(Qs,bb:data(),n)
  MaskApi._rlesFree(Qs,n)
  return bb
end

MaskApi.frBbox = function( bbs, h, w )
  if bbs:dim()==1 then bbs=bbs:view(1,bbs:size(1)) end
  local n, k = bbs:size(1), bbs:size(2); assert(k==4)
  local data = bbs:type('torch.DoubleTensor'):contiguous():data()
  local Qs = MaskApi._rlesInit(n)
  libmaskapi.rleFrBbox(Qs[0],data,h,w,n)
  return MaskApi._rlesToLua(Qs,n)
end

MaskApi.frPoly = function( poly, h, w )
  local n = #poly
  local Qs, Q = MaskApi._rlesInit(n), MaskApi._rlesInit(1)
  for i,p in pairs(poly) do
    local xy = p:type('torch.DoubleTensor'):contiguous():data()
    libmaskapi.rleFrPoly(Qs[i-1],xy,p:size(1)/2,h,w)
  end
  libmaskapi.rleMerge(Qs,Q[0],n,0)
  MaskApi._rlesFree(Qs,n)
  return MaskApi._rlesToLua(Q,1)[1]
end

MaskApi.drawMasks = function( img, masks, maxn, alpha, clrs )
  local n, h, w = masks:size(1), masks:size(2), masks:size(3)
  if not maxn then maxn=n end
  if not alpha then alpha=.4 end
  if not clrs then clrs=torch.rand(n,3)*1.4 end
  local out = img:clone():contiguous():float()
  for i=1,math.min(maxn,n) do
    local M = masks[i]:contiguous():data()
    local B = torch.ByteTensor(h,w):zero():contiguous():data()
    -- get boundaries B in masks M quickly
    for y=0,h-2 do for x=0,w-2 do
      local k=y*w+x
      if M[k]~=M[k+1] then B[k],B[k+1]=1,1 end
      if M[k]~=M[k+w] then B[k],B[k+w]=1,1 end
      if M[k]~=M[k+1+w] then B[k],B[k+1+w]=1,1 end
    end end
    -- softly embed masks into image and add solid boundaries
    for j=1,3 do
      local O,c,a = out[j]:data(), math.min(clrs[i][j],1), alpha
      for k=0,w*h-1 do if M[k]==1 then O[k]=O[k]*(1-a)+c*a end end
      for k=0,w*h-1 do if B[k]==1 then O[k]=c end end
    end
  end
  return out
end

--------------------------------------------------------------------------------

MaskApi._rlesToLua = function( Qs, n )
  local h, w, Rs = tonumber(Qs[0].h), tonumber(Qs[0].w), {}
  for i=1,n do Rs[i]={size={h,w}, counts={}} end
  for i=1,n do
    local s = libmaskapi.rleToString(Qs[i-1])
    Rs[i].counts=ffi.string(s)
    ffi.C.free(s)
  end
  MaskApi._rlesFree(Qs,n)
  return Rs
end

MaskApi._rlesFrLua = function( Rs )
  if #Rs==0 then Rs={Rs} end
  local n, h, w = #Rs, Rs[1].size[1], Rs[1].size[2]
  local Qs = MaskApi._rlesInit(n)
  for i=1,n do
    local c = Rs[i].counts
    if( torch.type(c)=='string' ) then
      local s=ffi.new("char[?]",#c+1); ffi.copy(s,c)
      libmaskapi.rleFrString(Qs[i-1],s,h,w)
    elseif( torch.type(c)=='torch.IntTensor' ) then
      libmaskapi.rleInit(Qs[i-1],h,w,c:size(1),c:contiguous():data())
    else
      assert(false,"invalid RLE")
    end
  end
  return Qs, n, h, w
end

MaskApi._rlesInit = function( n )
  local Qs = ffi.new("RLE[?]",n)
  for i=1,n do libmaskapi.rleInit(Qs[i-1],0,0,0,NULL) end
  return Qs
end

MaskApi._rlesFree = function( Qs, n )
  for i=1,n do libmaskapi.rleFree(Qs[i-1]) end
end

--------------------------------------------------------------------------------

ffi.cdef[[
  void free(void *ptr);
  typedef unsigned int uint;
  typedef unsigned long siz;
  typedef unsigned char byte;
  typedef double* BB;
  typedef struct { siz h, w, m; uint *cnts; } RLE;
  void rleInit( RLE *R, siz h, siz w, siz m, uint *cnts );
  void rleFree( RLE *R );
  void rlesInit( RLE **R, siz n );
  void rlesFree( RLE **R, siz n );
  void rleEncode( RLE *R, const byte *mask, siz h, siz w, siz n );
  void rleDecode( const RLE *R, byte *mask, siz n );
  void rleMerge( const RLE *R, RLE *M, siz n, int intersect );
  void rleArea( const RLE *R, siz n, uint *a );
  void rleIou( RLE *dt, RLE *gt, siz m, siz n, byte *iscrowd, double *o );
  void bbIou( BB dt, BB gt, siz m, siz n, byte *iscrowd, double *o );
  void rleToBbox( const RLE *R, BB bb, siz n );
  void rleFrBbox( RLE *R, const BB bb, siz h, siz w, siz n );
  void rleFrPoly( RLE *R, const double *xy, siz k, siz h, siz w );
  char* rleToString( const RLE *R );
  void rleFrString( RLE *R, char *s, siz h, siz w );
]]
